src/security/Handshake.cc

```diff
@@ -274,24 +274,30 @@ Security::HandshakeParser::parseModernRecord()
     Must(record.fragment.length() || record.type == ContentType::ctApplicationData);
 
     if (currentContentType != record.type) {
+        tkMessages.expectMore(false);
         parseMessages();
         Must(tkMessages.atEnd()); // no currentContentType leftovers
-        fragments = record.fragment;
         currentContentType = record.type;
-    } else {
-        fragments.append(record.fragment);
     }
 
-    if (tkRecords.atEnd() && !done)
-        parseMessages();
+    const auto haveUnparsedRecordBytes = !tkRecords.atEnd();
+    const auto expectMoreRecordLayerBytes = tkRecords.expectMore();
+    // TODO: consider adding BinaryTokenizer::exhausted() instead
+    const auto expectMoreMessageLayerBytes = haveUnparsedRecordBytes || expectMoreRecordLayerBytes;
+
+    tkMessages.expectMore(expectMoreMessageLayerBytes);
+    tkMessages.append(record.fragment);
+
+    parseMessages();
 }
 
 /// parses one or more "higher-level protocol" frames of currentContentType
 void
 Security::HandshakeParser::parseMessages()
 {
-    tkMessages.reset(fragments, false);
-    for (; !tkMessages.atEnd(); tkMessages.commit()) {
```

Comment:
> I suggest removing this atEnd() check and adjusting skipMessage() to throw InsufficientInput 

I followed your suggestion but faced a problem: parseMessages() started to call individual parsers (and IsufficientInput()) in cases where previously it just checked atEnd() and exited. This happens, for example, when the initial ClientHello comes - as a result, the transaction gets stuck. I tried a workaround (e.g., to check that currentContentType is not 0 before calling parseMessages()) - but it did not work in other cases. I left the atEnd() check for now.

Reply:
> to check that currentContentType is not 0

I made this fix first but then realized that we do not need that conditional parseMessages() at all.I tried to explain this logic in c5f70c1 description.

Reply:
Ok, I expected that we should change that old Must() with something more meaningful.

> correctly address zero currentContentType problem

It should have been addressed already because we removed parseMessages() from this block.